spec/re2/set_spec.rb

```diff
@@ -0,0 +1,121 @@
+      STDERR.reopen(File::NULL)
+    else
+      platform = RUBY_PLATFORM == 'java' ? RbConfig::CONFIG['host_os'] : RUBY_PLATFORM
+
+      case platform
+      when /mswin|mingw/i
+        STDERR.reopen('NUL')
+      when /amiga/i
+        STDERR.reopen('NIL')
+      when /openvms/i
+        STDERR.reopen('NL:')
+      else
+        STDERR.reopen('/dev/null')
+      end
+    end
+
+    yield
+  ensure
+    STDERR.reopen(original_stream)
+  end
+
+  describe "#initialize" do
+    it "returns an instance given no args" do
+      set = RE2::Set.new
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given only an anchor of :unanchored" do
+      set = RE2::Set.new(:unanchored)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given only an anchor of :anchor_start" do
+      set = RE2::Set.new(:anchor_start)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given only an anchor of :anchor_both" do
+      set = RE2::Set.new(:anchor_both)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "returns an instance given an anchor and options" do
+      set = RE2::Set.new(:unanchored, :case_sensitive => false)
+      expect(set).to be_a(RE2::Set)
+    end
+
+    it "raises an error if given an inappropriate type" do
+      expect { RE2::Set.new(0) }.to raise_error(TypeError)
+    end
+
+    it "raises an error if given an invalid anchor" do
+      expect { RE2::Set.new(:not_a_valid_anchor) }.to raise_error(
+        ArgumentError,
+        "anchor should be one of: :unanchored, :anchor_start, :anchor_both"
+      )
+    end
+  end
+
+  describe "#add" do
+    it "allows multiple patterns to be added", :aggregate_failures do
+      set = RE2::Set.new
+      expect(set.add("abc")).to eq(0)
+      expect(set.add("def")).to eq(1)
+      expect(set.add("ghi")).to eq(2)
+    end
+
+    it "rejects invalid patterns when added" do
+      set = RE2::Set.new(:unanchored, :log_errors => false)
+      expect { set.add("???") }.to raise_error(ArgumentError, /str rejected by RE2::Set->Add()/)
+    end
+
+    it "raises an error if called after #compile" do
+      set = RE2::Set.new(:unanchored, :log_errors => false)
+      set.add("abc")
+      set.compile
+      silence_stderr do
+        expect { set.add("def") }.to raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe "#compile" do
+    it "compiles the set without error" do
+      set = RE2::Set.new
+      set.add("abc")
+      set.add("def")
+      set.add("ghi")
+      expect(set.compile).to be_truthy
+    end
+  end
+
+  describe "#match" do
+    it "matches against multiple patterns" do
+      set = RE2::Set.new
+      set.add("abc")
+      set.add("def")
+      set.add("ghi")
+      set.compile
+      expect(set.match("abcdefghi")).to eq([0, 1, 2])
+    end
+
+    it "raises an error if called before #compile" do
+      pending "Versions of re2 with ABI 0 will only return false here"
```

Comment:
This behaviour discrepancy between underlying versions of re2 makes me uneasy: if you have a modern version of the library, `#match` will raise an exception but if you have an old version, you'll get a `false` return value and no error.

My concern here is that a change in the underlying re2 library will cause significant change in any client programs, especially as a raised exception will alter the control flow.

However, it doesn't seem trivial to backfill this behaviour to re2 with an ABI version of 0 as not enough information is exposed by `RE2::Set::Match()` there.

Reply:
I've updated this in c00eda1385c722b1063a7d33d26cb6e38b06f4b8  so there is now a `RE2::Set.match_raises_errors?` predicate that returns `true` or `false` based on the underlying re2 library's support. We can then distinguish the two behaviours: raises errors or simply returning an empty array of matches.

I'm still not decided on the final API here though. Should calling `RE2::Set#match` raise or not raise errors depending on the value of `RE2::Set.match_raises_errors?` or should we take a leaf from [`Integer`](https://ruby-doc.org/core-3.1.2/Kernel.html#Integer-method) and have an `exception` argument that defaults to `true` and raises a [`NotImplementedError`](https://ruby-doc.org/core-3.1.2/NotImplementedError.html) if the user tries to use that against a version of re2 that doesn't output error information?

```ruby
set = RE2::Set.new
set.add("foo")
set.add("bar")
set.match("foobar")
# NotImplementedError "current version of RE2::Set::Match does not output error information"
set.match("foobar", exception: false)
# => []
```

Reply:
I implemented the `:exception` option in ebfe0336dc6fed98496e8141aa992ad0e6a2dbb0.